cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/DirectRuntimeRequestValidator.java

```diff
@@ -73,48 +73,49 @@
     long pollTimeMillis = cConf.getLong(Constants.RuntimeMonitor.POLL_TIME_MS);
     this.programRunsCache = CacheBuilder.newBuilder()
       .expireAfterWrite(pollTimeMillis, TimeUnit.MILLISECONDS)
-      .build(new CacheLoader<ProgramRunId, Boolean>() {
+      .build(new CacheLoader<ProgramRunId, RunRecordDetail>() {
         @Override
-        public Boolean load(ProgramRunId programRunId) throws IOException, UnauthorizedException {
-          return isValid(programRunId);
+        public RunRecordDetail load(ProgramRunId programRunId) throws UnauthorizedException, NotFoundException,
+          IOException {
+          return runRecord(programRunId);
         }
       });
   }
 
   @Override
-  public void validate(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
+  public RunRecordDetail checkRunRecord(ProgramRunId programRunId, HttpRequest request) throws BadRequestException {
     accessEnforcer.enforce(programRunId, authenticationContext.getPrincipal(), StandardPermission.GET);
-    boolean exists;
+    RunRecordDetail runRecord;
     try {
-      exists = programRunsCache.get(programRunId);
+      runRecord = programRunsCache.get(programRunId);
+      return runRecord;
     } catch (Exception e) {
       throw new ServiceUnavailableException(Constants.Service.RUNTIME, e);
     }
-    if (!exists) {
-      throw new BadRequestException("Program run " + programRunId + " is not valid");
-    }
   }
 
-  /**
-   * Checks if the given {@link ProgramRunId} is valid.
-   */
-  private boolean isValid(ProgramRunId programRunId) throws IOException, UnauthorizedException {
+  private RunRecordDetail runRecord(ProgramRunId programRunId) throws IOException, UnauthorizedException,
+    NotFoundException {
     RunRecordDetail runRecord = TransactionRunners.run(txRunner, context -> {
       return AppMetadataStore.create(context).getRun(programRunId);
     }, IOException.class);
 
     if (runRecord != null) {
-      return !runRecord.getStatus().isEndState();
+      return runRecord;
```

Comment:
What is the impact if we remove this check? Originally this method was being used to validate the record status.